src/particles/beam/BeamParticleContainerInit.cpp

```diff
@@ -463,6 +463,214 @@ InitBeamFixedWeightSlice (int slice, int which_slice)
+            const amrex::Real local_weight = 0.5_rt*(pdf_zmin+pdf_zmax);
+
+            const amrex::Real mean_particles = num_to_add_now*local_weight/integral;
+
+            if (mean_particles >= 0) {
+                const unsigned int n = amrex::RandomPoisson(mean_particles);
+                m_num_particles_slice[slice] += n;
+                num_added += n;
+            } else {
+                const unsigned int n = std::min(amrex::RandomPoisson(-mean_particles),
+                                                m_num_particles_slice[slice]);
+                m_num_particles_slice[slice] -= n;
+                num_added -= n;
+            }
+        }
+    }
+}
+
+void
+BeamParticleContainer::
+InitBeamFixedWeightPDFSlice (int slice, int which_slice)
+{
+    HIPACE_PROFILE("BeamParticleContainer::InitBeamFixedWeightPDFSlice()");
+    using namespace amrex::literals;
+
+    if (!Hipace::HeadRank() || m_num_particles == 0) { return; }
+
+    unsigned int num_to_add_full = 0;
+    for (int r=m_pdf_ref_ratio-1; r>=0; --r) {
+        num_to_add_full += m_num_particles_slice[slice*m_pdf_ref_ratio+r];
+    }
+    if (m_do_symmetrize) {
+        resize(which_slice, 4*num_to_add_full, 0);
+    } else {
+        resize(which_slice, num_to_add_full, 0);
+    }
+
+    unsigned int loc_index = 0;
+    for (int r=m_pdf_ref_ratio-1; r>=0; --r) {
+        const unsigned int num_to_add = m_num_particles_slice[slice*m_pdf_ref_ratio+r];
+        if (num_to_add == 0) continue;
+
+        auto& particle_tile = getBeamSlice(which_slice);
+        // Access particles' SoA
+        amrex::GpuArray<amrex::ParticleReal*, BeamIdx::real_nattribs> rarrdata =
+            particle_tile.GetStructOfArrays().realarray();
+        amrex::GpuArray<int*, BeamIdx::int_nattribs> iarrdata =
+            particle_tile.GetStructOfArrays().intarray();
+
+        const uint64_t pid = m_id64;
+        m_id64 += num_to_add;
+
+        const amrex::Real clight = get_phys_const().c;
+        const bool do_symmetrize = m_do_symmetrize;
+        const bool peak_density_is_specified = m_peak_density_is_specified;
+        const amrex::Real z_foc = m_z_foc;
+        const amrex::Real radius = m_radius;
+        const amrex::Real weight = m_total_weight / m_num_particles;
+        const auto pos_func = m_pdf_pos_func;
+        const auto u_func = m_pdf_u_func;
+        const amrex::Geometry& geom = Hipace::GetInstance().m_3D_geom[0];
+        const amrex::Real dz = geom.CellSize(2) / m_pdf_ref_ratio;
+        const amrex::Real zmin = geom.ProbLo(2) + (slice*m_pdf_ref_ratio+r)*dz;
+        const amrex::Real zmax = geom.ProbLo(2) + (slice*m_pdf_ref_ratio+r+1)*dz;
+        const amrex::Real lo_weight = m_pdf_func(zmin);
+        const amrex::Real hi_weight = m_pdf_func(zmax);
+        AMREX_ALWAYS_ASSERT(lo_weight + hi_weight > 0._rt);
+        const bool use_taylor = std::min(lo_weight, hi_weight)*1.1 > std::max(lo_weight, hi_weight);
```

Comment:
Could you add some comments of how `use_taylor` is set, and what it changes in the behavior?